Design/WorkspacePropertyWithIndex.md

```diff
@@ -0,0 +1,95 @@
+  - Some algorithms do not validate for duplicate spectrum numbers and double-process the corresponding spectra (see [16651](https://github.com/mantidproject/mantid/issues/16651)).
+  - Validations are not unit-tested to a good extent.
+
+## Design Goals
+ - Provide a property type which can specify a workspace and allowed (and predefined) input index types.
+ - Unified behaviour on error (when bad indices are provided).
+ - Hide index translations from the developer (this will take place internally within the property).
+ - Automatically generate the required GUI input controls based on the specified index types.
+ - Return the workspace and correct index set simultaneously in one call to `PropertyManager::getProperty`. 
+ 
+## Implementation
+
+### Basics
+
+- Extend the `WorkspaceProperty` type.
+- Own an `ArrayProperty` which can make use of `ListPropertyWidget` for capturing list data from the GUI.
+- Add a set of flags which control allowed index types (for use in the constructor).
+
+### Details
+
+1. The `WorkspacePropertyWithIndex` type constructor will accept the following flags (or combinations of flags):
+	- `IndexType::SpectrumNumber` for list of global<sup>1</sup> spectrum numbers typically from 1-N.
+	- `IndexType::WorkspaceIndex` for list of global workspace indices typically from 0-N.
+	- `IndexType::DetectorID` for list of detector ids within the instrument.
+2. Users will be restricted to entering only one of the allowed index types at the GUI level. 
+3. Validation will occur in `IndexInfo::makeIndexSet`(which throws for bad indices), not in every algorithm.
+4. The `ArrayProperty` will be exposed e.g `WorkspacePropertyWithIndex::getArrayProperty` in order to make use of `ListPropertyWidget`.
+5. Overload function call operator to return `std::pair`.
+6. `IndexInfo` will be used to translate the indices at the point of access i.e. `WorkspacePropertyWithIndex::operator()()`.
+7. There must be a few modifications to the GUI generation code in order to support this type:
+	-  `PropertyWidgetFactory` may need to handle `WorkspacePropertyWithIndex` specifically.
+	-  May need to introduce a concept of a group of property widgets.
+
+<sup>1</sup><sub>The distinction is made here between local (a subset of all data partitioned on an MPI Rank) and global (entire data set).</sub>
+
+### Example usage
+
+**Property Declaration**:
+
+```cpp
+//Property name is fixed
+declareProperty(make_unique<WorkspacePropertyWithIndex>(IndexType::SpectrumNumber|IndexType::WorkspaceIndex));
+```  
+
+**Algorithm Dialog Box (GUI)**:
+
+**InputWorkspaceWithIndex** [ _________________ ]<br>
+( **.** ) **SpectrumNumber** (  ) **WorkspaceIndex**<br>
+**Indices** [ _________________ ]
+ 
+**Client Code Access**:
+
+C++:
+
+```cpp
+//C++11
+MatrixWorkspace_const_sptr inputWS;
+IndexSet indices;
+std::tie(inputWS, indices) = getProperty("InputWorkspaceWithIndex");
+```
+
+Python:
+
+```python
+inputWS, indices = getProperty("InputWorkspaceWithIndex")
+```
```

Comment:
Could you add a discussion of how an algorithm with such a property would be called? What would you pass to `setProperty` in C++? In Python, would there be a single argument that accepts a compount argument containing the workspace and this indices?

Even if you do not have a solution ready, putting something in here aids the discussion.